Assets/MRTK/SDK/Experimental/Drawstring/Scripts/ElasticSystem.cs

```diff
@@ -0,0 +1,196 @@
+    {
+        public LinearElasticSystem(float initialValue, float initialVelocity,
+                                ElasticExtentProperties<float> extentInfo, ElasticProperties elasticProperties)
+        : base(initialValue, initialVelocity, extentInfo, elasticProperties) { }
+
+
+        /// <summary>
+        /// Update the internal state of the damped harmonic oscillator, given the forcing/desired value.
+        /// </summary>
+        /// <param name="forcingValue">Input value, for example, a desired manipulation position.</param>
+        /// <param name="deltaTime">Amount of time that has passed since the last update.</param>
+        public override float ComputeIteration(float forcingValue, float deltaTime)
+        {
+            // F = -kx - (drag * v)
+            var force = (forcingValue - currentValue) * elasticProperties.hand_k - elasticProperties.drag * currentVelocity;
+
+            // Distance that the current stretch value is from the end limit.
+            float distFromEnd = extentInfo.maxStretch - currentValue;
+
+            // If we are extended beyond the end cap,
+            // add one-sided force back to the center.
+            if (currentValue > extentInfo.maxStretch)
+            {
+                force += distFromEnd * elasticProperties.end_k;
+            }
+            else
+            {
+                // Otherwise, add standard bidirectional magnetic/snapping force towards the end marker. (optional)
+                if (extentInfo.snapToMax)
+                {
+                    force += (distFromEnd) * elasticProperties.end_k * (1.0f - Mathf.Clamp01(Mathf.Abs(distFromEnd / elasticProperties.snap_radius)));
+                }
+            }
+
+            distFromEnd = extentInfo.minStretch - currentValue;
+            if (currentValue < extentInfo.minStretch)
+            {
+                force += distFromEnd * elasticProperties.end_k;
+            }
+            else
+            {
+                // Otherwise, add standard bidirectional magnetic/snapping force towards the end marker. (optional)
+                if (extentInfo.snapToMax)
+                {
+                    force += (distFromEnd) * elasticProperties.end_k * (1.0f - Mathf.Clamp01(Mathf.Abs(distFromEnd / elasticProperties.snap_radius)));
+                }
+            }
+
+            // Iterate over each snapping point, and apply forces as necessary.
+            foreach (float snappingPoint in extentInfo.snapPoints)
+            {
+                // Calculate distance from snapping point.
+                var distFromSnappingPoint = snappingPoint - currentValue;
+
+                // Snap force is calculated by multiplying the "-kx" factor by
+                // a clamped distance factor. This results in an overall
+                // hyperbolic profile to the force imparted by the snap point.
+                force += (distFromSnappingPoint) * elasticProperties.snap_k
```

Comment:
If I understand this correct this ends up adding a force for each snap point, not just the snap point(s) that are closest on either side right?)

Reply:
Yes.

Reply:
For multiple snap points nearby each other, this is necessary to accurately model the forces involved.